File: backend/src/main/kotlin/com/alpha/service/CategoryService.kt
package com.alpha.service

import com.alpha.domain.entity.CategoryEntity
import com.alpha.domain.repository.CategoryRepository
import com.alpha.service.dto.CategoryResponse
import com.alpha.service.dto.CreateCategoryRequest
import com.alpha.service.dto.UpdateCategoryRequest
import com.alpha.service.exception.NotFoundException
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
class CategoryService(
    private val categoryRepository: CategoryRepository
) {
    
    fun getAllCategories(): List<CategoryResponse> {
        return categoryRepository.findRootCategories()
            .map { it.toResponse() }
    }
    
    fun getCategory(id: UUID): CategoryResponse {
        val category = categoryRepository.findById(id)
            .orElseThrow { NotFoundException("Category not found") }
        return category.toResponse()
    }
    
    @Transactional
    fun createCategory(request: CreateCategoryRequest): CategoryResponse {
        val category = CategoryEntity().apply {
            name = request.name
            slug = request.name.lowercase().replace(" ", "-")
            description = request.description
            icon = request.icon
            sortOrder = request.sortOrder
            isActive = request.isActive
        }
        return categoryRepository.save(category).toResponse()
    }
    
    @Transactional
    fun updateCategory(id: UUID, request: UpdateCategoryRequest): CategoryResponse {
        val category = categoryRepository.findById(id)
            .orElseThrow { NotFoundException("Category not found") }
        
        request.name?.let { category.name = it }
        request.description?.let { category.description = it }
        request.icon?.let { category.icon = it }
        request.sortOrder?.let { category.sortOrder = it }
        request.isActive?.let { category.isActive = it }
        
        return categoryRepository.save(category).toResponse()
    }
    
    @Transactional
    fun deleteCategory(id: UUID) {
        if (!categoryRepository.existsById(id)) {
            throw NotFoundException("Category not found")
        }
        categoryRepository.deleteById(id)
    }
    
    private fun CategoryEntity.toResponse(): CategoryResponse {
        return CategoryResponse(
            id = id!!,
            name = name,
            slug = slug,
            description = description,
            icon = icon,
            parentId = parent?.id,
            sortOrder = sortOrder,
            isActive = isActive,
            createdAt = createdAt!!
        )
    }
}

File: backend-scala-zio/src/main/scala/com/alpha/service/CategoryService.scala
package com.alpha.service

import zio.*
import com.alpha.repository.*
import com.alpha.domain.model.*
import com.alpha.dto.*
import com.alpha.util.SlugGenerator
import com.alpha.provider.*
import java.util.UUID

trait CategoryService:
  def getAllCategories: Task[List[Category]]
  def getCategory(id: UUID): Task[Option[Category]]
  def createCategory(req: CreateCategoryRequest): Task[Category]
  def updateCategory(id: UUID, req: UpdateCategoryRequest): Task[Category]
  def deleteCategory(id: UUID): Task[Unit]

object CategoryService:
  val layer: ZLayer[CategoryRepository & TimeProvider & UUIDProvider, Nothing, CategoryService] =
    ZLayer.fromFunction(new CategoryServiceImpl(_, _, _))

class CategoryServiceImpl(categoryRepo: CategoryRepository, timeProvider: TimeProvider, uuidProvider: UUIDProvider)
  extends CategoryService:

  override def getAllCategories: Task[List[Category]] =
    categoryRepo.findAll

  override def getCategory(id: UUID): Task[Option[Category]] =
    categoryRepo.findById(id)

  override def createCategory(req: CreateCategoryRequest): Task[Category] =
    val baseSlug                                                 = SlugGenerator.generate(req.name)
    def findUniqueSlug(slug: String, attempt: Int): Task[String] =
      categoryRepo.existsBySlug(slug).flatMap {
        case true if attempt < 10 => findUniqueSlug(s"$slug-$attempt", attempt + 1)
        case true                 => ZIO.succeed(s"$slug-${java.lang.System.currentTimeMillis()}")
        case false                => ZIO.succeed(slug)
      }
    for
      slug    <- findUniqueSlug(baseSlug, 1)
      category = Category(
                   id = uuidProvider.randomUUID(),
                   name = req.name,
                   slug = slug,
                   description = req.description,
                   icon = req.icon,
                   parentId = req.parentId,
                   sortOrder = req.sortOrder,
                   isActive = true,
                   createdAt = timeProvider.now(),
                   updatedAt = None
                 )
      _       <- categoryRepo.create(category)
    yield category

  override def updateCategory(id: UUID, req: UpdateCategoryRequest): Task[Category] =
    for
      categoryOpt <- categoryRepo.findById(id)
      category    <- ZIO.fromOption(categoryOpt).orElseFail(new Exception("Category not found"))
      updated      = category.copy(
                       name = req.name.getOrElse(category.name),
                       description = req.description.orElse(category.description),
                       icon = req.icon.orElse(category.icon),
                       parentId = req.parentId.orElse(category.parentId),
                       sortOrder = req.sortOrder.getOrElse(category.sortOrder),
                       isActive = req.isActive.getOrElse(category.isActive),
                       updatedAt = Some(timeProvider.now())
                     )
      _           <- categoryRepo.update(updated)
    yield updated

  override def deleteCategory(id: UUID): Task[Unit] =
    categoryRepo.delete(id).unit
